### DISCUSSION TITLE
Micropython and typing

### DISCUSSION THREAD

**Discussion by brianreinhold** [micropython/micropython#13711] [General]
I really dislike the run time typing used by Python, Micropython, and Typescript (and perhaps other languages). When  a project becomes large and creates objects with complex fields and structures and has to be understood by someone else, it becomes a nightmare. I don't want to put anyone else through that agony so I strictly type my code.

However, I have run into situations where it appears the typing causes a run time syntax error. For example, this line of code declaring a bunch of variables
```conn_handle, start_handle, end_handle, uuid = data```
runs fine on the PICO-W. But if I type it
```conn_handle: int, start_handle: int, end_handle: int, uuid = data```

it gives a syntax error.

Why?

**Comment by andrewleech** [comment]
Careful, you risk starting a flame war with those kind of statements ;-) 
https://erik-engheim.medium.com/the-many-advantages-of-dynamic-languages-267d08f4c7

Splitting a tuple/list/etc out into separate variables is called "unpacking". Type hinting unpacking this way doesn't work in python either, it needs to be declared in a different way. A few suggestions have been made here: https://stackoverflow.com/questions/52082939/type-hints-when-unpacking-a-tuple

**Reply by brianreinhold** [reply]
I guess its because I am not enough of a programming whiz (certainly not in Python) to figure out how things work without typing so I want to type everything. I will personally be glad if I have to look back on my code after a few months reprieve.  But I don't want to cause a flame war, especially with you as you have been a great help. There are things I wish I could type but cant (like iterations)
```
for some_variable in variables:
```
There is no way to type `some_variables` which may come from one of the dunder functions and could be a sub-element of `variables`.

IN any case, would a way to type such a list of variables that gave me problems be:
```
conn_handle: int
start_handle: int
end_handle: int
uuid: int | str | bytes = data
```

**Reply by Josverl** [reply]
you can also type the complex object, most static typecheckers will be able to follow that though into the variables that it is assigned to:
(and you can use that as a returned type on any methods you define in python, or in a .pyi stub file for any C-Module.

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqArggDYCmANFAKopJgrUjkBu5AhqQPrwLkAoAQDEwREMTLkoUALwzJFANqoY1VepRra9Riq2UAzjBCUARnBjlDAXRsCAJhxgcAXFFHjF0%2BQAoAjNQATNQAzNQARAAsAKwAbADsEQCUQgDGjCjcABYcKA4U1MYceDl5BVRQ5Pll%2BYVQRERIDnJQTi5QQizsXLyI5L4ZKFm5deSp3Zw8fAPVDrUVE2xTffy%2Bjc2pAkA)

```python
from typing import Tuple, Union, reveal_type

FourTuple  =  Tuple[int, int, int, Union[int,str,bytes]]
data: FourTuple = (1, 2, 3, "4567")

conn_handle, start_handle, end_handle, uuid = data 

reveal_type(conn_handle)
reveal_type(end_handle)
reveal_type(uuid)
```

**Reply by GitHubsSilverBullet** [reply]
Instead of extensive typing, maybe the assert approach would be useful.
```python
a =  [1,2,3,4,4.5]
for x in a:
    assert type(x) is int, f'{x} is not int'

AssertionError: 4.5 is not int
```

**Reply by Josverl** [reply]
**_Note_**: asserts are generally dropped when cross-compiling with -o2, which is common to optimize for speed&memory.

**Reply by Josverl** [reply]
_FYI_ The Python typing community hangs out on : https://github.com/python/typing/discussions.  
They have been helpful but may have complex answers to seemingly simple questions.

If you have a more elaborate example that you want some eyes on, feel free to tag me or post a snippet in a Gist

**Reply by brianreinhold** [reply]
@GitHubsSilverBullet I am primarily interested in typing to help others (and myself down the road) understand what my code is doing. It deals with Bluetooth which requires not only understanding of what I am trying to do but the Bluetooth spec. When all the variables are typed intellisense becomes a very helpful tool and is a great aid in not only writing code but understanding it.
